### PR TITLE
Indexing coordinator

### PR DESCRIPTION
**Pull request by umbreak**

### CODE REVIEW

File: delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/indexing/ViewLens.scala [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+import ch.epfl.bluebrain.nexus.sourcing.projections.ProjectionId.ViewProjectionId
+
+import java.util.UUID
+
+/**
+  * A lens on a View
+  * @tparam A the view type
+  */
+trait ViewLens[A] {
```
Comment (imsdu) [inline delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/indexing/ViewLens.scala:11]: Should not we just have a trait defined in the sdk that would be needed to be implemented by the different views in the different plugins rather than having too defined a Lens ?
Plus we only need basic information here

**Review by imsdu — CHANGES_REQUESTED** [review]

File: delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/indexing/ViewLens.scala (hunk 1 above)
Comment (umbreak) [reply delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/indexing/ViewLens.scala:11]: I didn't want to couple those things _yet_.
Maybe we end up doing it this way, maybe not.

File: delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/indexing/IndexingState.scala [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package ch.epfl.bluebrain.nexus.delta.sdk.indexing
+
+import ch.epfl.bluebrain.nexus.sourcing.projections.stream.StreamSupervisor
+
+sealed private[indexing] trait IndexingState extends Product with Serializable
+
+private[indexing] object IndexingState {
+  final case object Initial                                         extends IndexingState
+  final case class Current(rev: Long, supervisor: StreamSupervisor) extends IndexingState
```
Comment (imsdu) [inline delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/indexing/IndexingState.scala:9]: It may be another PR but it would be nice to know when it has been started for example

File: delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/indexing/IndexingState.scala [hunk 3]
```diff
@@ -0,0 +1,10 @@
+package ch.epfl.bluebrain.nexus.delta.sdk.indexing
+
+import ch.epfl.bluebrain.nexus.sourcing.projections.stream.StreamSupervisor
+
+sealed private[indexing] trait IndexingState extends Product with Serializable
+
+private[indexing] object IndexingState {
+  final case object Initial                                         extends IndexingState
```
Comment (imsdu) [inline delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/indexing/IndexingState.scala:8]: We could have a different state for Stop to make a distinction between views that have never been started and the ones that ran at least once

File: delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/indexing/IndexingStreamCoordinator.scala [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+  def apply[V: ViewLens](
+      entityType: String,
+      buildStream: BuildStream[V],
+      projection: Projection[Unit],
+      config: EventSourceProcessorConfig,
+      retryStrategy: RetryStrategy[Throwable]
+  )(implicit V: ClassTag[V], as: ActorSystem[Nothing], sc: Scheduler): Task[IndexingStreamCoordinator[V]] = {
+
+    def start(view: V): Task[IndexingState] = {
+      val stream = projection.progress(view.projectionId).flatMap(buildStream(view, _))
```
Comment (wwajerowicz) [inline delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/indexing/IndexingStreamCoordinator.scala:57]: We also need an option to restart indexing the view from the beginning for the cases when we just want to reindex a view without updating it.

**Review by imsdu — CHANGES_REQUESTED** [review]

**Review by wwajerowicz — CHANGES_REQUESTED** [review]

File: delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/indexing/IndexingState.scala (hunk 2 above)
Comment (umbreak) [reply delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/indexing/IndexingState.scala:9]: This information is already present in the progress offset, which as per your request is handled separately in a distributed data cache

File: delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/indexing/IndexingState.scala (hunk 3 above)
Comment (umbreak) [reply delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/indexing/IndexingState.scala:8]: Again, this information is going to be obtained using the progress

**Review by bogdanromanx — DISMISSED** [review]

File: delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/indexing/IndexingStreamCoordinator.scala [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+object IndexingStreamCoordinator {
+
+  type BuildStream[V] = (V, ProjectionProgress[Unit]) => Task[Stream[Task, Unit]]
+
+  private[indexing] type Agg = Aggregate[String, IndexingState, IndexingCommand, IndexingState, Throwable]
+
+  /**
+    * Construct a [[IndexingStreamCoordinator]] relying on the underlying transient [[ShardedAggregate]]
+    */
+  def apply[V: ViewLens](
```
[the remark this answers is not in the document]
Comment (umbreak) [reply delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/indexing/IndexingStreamCoordinator.scala:48]: If we need extra work to use cluster sharding, we don't want to use it. The balancing benefits do not outperform the extra hussle.

[the remark this answers is not in the document]
Comment (bogdanromanx) [reply delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/indexing/IndexingStreamCoordinator.scala:48]: all cluster singletons run on the oldest node; this affects the performance of the entire system... but let's discuss in #2031 if you're not confident this would be a good change.

File: delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/indexing/IndexingState.scala (hunk 2 above)
Comment (imsdu) [reply delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/indexing/IndexingState.scala:9]: I meant when the current StreamSupervisor started, this part is not in progress offset.

**Review by wwajerowicz — DISMISSED** [review]

File: delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/indexing/IndexingState.scala (hunk 2 above)
Comment (umbreak) [reply delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/indexing/IndexingState.scala:9]: We don't care about when it started, we care about the progress.

If we ever care about when it started, I will add it

**Review by wwajerowicz — APPROVED** [review]

**Review by bogdanromanx — APPROVED** [review]
